crates/shared/src/price_estimation/native/oneinch.rs

```diff
@@ -0,0 +1,237 @@
+    futures::{future::BoxFuture, FutureExt},
+    model::order::OrderKind,
+    number::nonzero::U256 as NonZeroU256,
+    primitive_types::{H160, U256},
+    reqwest::Client,
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+        time::Duration,
+    },
+};
+
+const BASE_URL: &str = "https://api.1inch.dev/";
+const REFRESH_INTERVAL: Duration = Duration::from_secs(12);
+
+struct OneInch {
+    client: Client,
+    chain_id: u64,
+    // Denominated in wei
+    prices: Arc<Mutex<HashMap<H160, U256>>>,
+    native_price_estimation_amount: NonZeroU256,
+    native_token: H160,
+}
+
+impl OneInch {
+    #[allow(dead_code)]
+    pub fn new(
+        client: Client,
+        chain_id: u64,
+        native_price_estimation_amount: NonZeroU256,
+        native_token: H160,
+    ) -> Self {
+        let instance = Self {
+            client,
+            chain_id,
+            prices: Arc::new(Mutex::new(HashMap::new())),
+            native_price_estimation_amount,
+            native_token,
+        };
+        instance.update_prices_in_background();
+        instance
+    }
+
+    fn update_prices_in_background(&self) {
+        let client = self.client.clone();
+        let chain = self.chain_id;
+        let prices = self.prices.clone();
+        tokio::task::spawn(async move {
+            loop {
+                match update_prices(&client, chain).await {
+                    Ok(new_prices) => {
+                        *prices.lock().unwrap() = new_prices;
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err);
+                    }
+                }
+                tokio::time::sleep(REFRESH_INTERVAL).await;
+            }
+        });
+    }
+}
+
+impl NativePriceEstimating for OneInch {
+    fn estimate_native_price<'a>(
+        &'a self,
+        token: &'a H160,
+    ) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let prices = self.prices.lock().unwrap();
+            let price = prices
+                .get(token)
+                .ok_or_else(|| PriceEstimationError::NoLiquidity)?;
+            Ok(price.to_f64_lossy() / 1e18)
+        }
+        .boxed()
+    }
+}
+
+impl PriceEstimating for OneInch {
+    fn estimate(&self, query: Arc<Query>) -> BoxFuture<'_, PriceEstimateResult> {
+        async move {
+            if query.kind != OrderKind::Buy
+                || query.in_amount != self.native_price_estimation_amount
+                || query.buy_token != self.native_token
+            {
+                return Err(PriceEstimationError::UnsupportedOrderType(
+                    "Non native price quote".to_string(),
+                ));
+            }
+
+            let prices = self.prices.lock().unwrap();
+            let price = prices
+                .get(&query.sell_token)
+                .ok_or_else(|| PriceEstimationError::NoLiquidity)?;
+            let reverse_price = 1. / (price.to_f64_lossy() / 1e18);
```

Comment:
This should probably use`checked_div` in case the API ever returns `0` for one of the prices.
Also you can already `drop` the `prices` lock before computing `reverse_price`.

Reply:
Nice catch!